lib/mongo/server/connection_pool.rb

```diff
@@ -513,40 +535,83 @@ def close_idle_sockets
         end
       end
 
+      # Stop the background populator thread and clean up any connections created
+      # which have not been connected yet.
+      #
+      # Used when closing the pool or when terminating the bg thread for testing
+      # purposes. In the latter case, this method must be called before the pool
+      # is used, to ensure no connections in pending_connections were created in-flow
+      # by the check_out method.
+      #
+      # @api private
+      def stop_populator
+        @populator.stop!
+
+        @lock.synchronize do
+          # If stop_populator is called while populate is running, there may be
+          # connections waiting to be connected, connections which have not yet
+          # been moved to available_connections, or connections moved to available_connections
+          # but not deleted from pending_connections. These should be cleaned up.
+          until @pending_connections.empty?
+            connection = @pending_connections.take(1).first
+            connection.disconnect!
+            @pending_connections.delete(connection)
+          end
+        end
+      end
+
       # Creates and adds a connection to the pool, if the pool's size is below
-      # min_size.
+      # min_size. Retries once if an error is encountered during this process
+      # and raises if a second error occurs.
       #
       # Used by the pool populator background thread.
       #
       # @return [ true | false ] Whether this method should be called again
       #   to create more connections.
+      # @raise [ Mongo::Error ] The second error raised if a retry occured
+      #
       # @api private
       def populate
         return false if closed?
 
-        @lock.synchronize do
-          if !closed? && unsynchronized_size < min_size
-            @available_connections << create_connection
-            true
-          else
-            false
-          end
+        begin
+          return create_and_add_connection
+        rescue Exception
+          # an error was encountered while connecting the connection,
+          # ignore this first error and try again.
+        end
+
+        begin
+          return create_and_add_connection
+        rescue Exception
```

Comment:
... but the semaphore signal should happen on all exceptions, regardless of whether they are retried. This is probably easiest to achieve by moving this rescue block to the top level of the method.